File: server/node/test/hz/message/mysql_post_test.js
#!/usr/bin/env node

'use strict';

const assert = require('assert');
const mysql_holder = require('../../../src/hz/message/mysql_holder.js');
const mysql_post = require('../../../src/hz/message/mysql_post.js');

// 数据库用户
const user = 'root';
// 数据库密码
const password = 'Ningning~1';
// 数据库名称
const database = 'test_message';

// 测试Post内容
const testPosts = [
	{'content': 'abcdefghijklmnopqrstuvwxyz'},
	{'content': 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'},
	{'content': '1234567890'},
	{'content': '红尘世界？一片雾茫茫！'},
	{'content': '<p>红尘世界？<em>一片雾茫茫！</em></p>'},
];

describe('mysql_post.MysqlPostTest', function() {
	beforeEach(function(done) {
		let _this = this;
		
		_this._hdr = new mysql_holder.MysqlHolder(user, password, database, function() {
			_this._pst = new mysql_post.MysqlPost(_this._hdr.inst(), function(err) {
				assert(!err);
				
				done();
			});
		});
	});

	afterEach(function(done) {
		let _this = this;
		
		_this._hdr.destroy(function() {});
		_this._hdr.close(function() {
			done();
		});
	});
	
	it('mysql_post.MysqlPostTest.getCount()', function(done) {
		let _this = this;
		
		this.timeout(0);
		
		// 测试初始post个数为0
		_this._pst.getCount(function(err, res) {
            assert(!err);
			assert.deepEqual(res, {status: 'ok', data: {'count': 0}});
		});
		
		// 添加测试post
		for (let i in testPosts) {
			_this._pst.add(testPosts[i], function() {});
		}
					
		// 测试post个数为添加的post个数
		_this._pst.getCount(function(err, res) {
            assert(!err);
			assert.deepEqual(res, {status: 'ok', data: {'count': testPosts.length}});
			
			done();
		});
	});
	
	it('mysql_post.MysqlPostTest.getAll()', function(done) {
		let _this = this;
		
		_this.timeout(0);
		
		// 测试初始的post为空
		_this._pst.getAll(function(err, res) {
            assert(!err);
			assert.deepEqual(res, {'status': 'ok', 'data': []});
		});
		
		// 添加测试post
		for (let i in testPosts) {
			_this._pst.add(testPosts[i], function() {});
		}
		
		// 测试获取的post时间递减，和添加的post顺序相反、content一致
		_this._pst.getAll(function(err, res) {
            assert(!err);
			assert.equal(res.status, 'ok');
			let posts = res.data;
			assert.equal(posts.length, testPosts.length);
			let savedTimestamp = Number.MAX_SAFE_INTEGER;
			for (let i in posts) {
				let post = posts[i];
				let timestamp = post.timestamp;
				assert(timestamp <= savedTimestamp);
				savedTimestamp = timestamp;
				let content = post.content;
				assert.equal(content, testPosts[testPosts.length - i - 1].content);
			}
			
			done();
		});
	});
	
	it('mysql_post.MysqlPostTest.getById()', function(done) {
		let _this = this;
		
		_this.timeout(0);
		
		// 测试对空的post，通过id获取错误
		_this._pst.getById({id: Number.MAX_SAFE_INTEGER}, function(err, res) {
            assert(!err);
			assert.equal(res.status, 'none_target');
		});
		
		// 添加测试post		
		for (let i in testPosts) {
			_this._pst.add(testPosts[i], function() {});
		}

		// 测试对每一个添加的post，和通过对应的id获取的post一致
		_this._pst.getAll(function(err, res) {
			let posts = res.data;
			for (let i in posts) {
				let post = posts[i];
				let id = post.id;
				_this._pst.getById({id: id}, function(ierr, ires) {
                    assert(!ierr);
					assert.deepEqual(ires, {status: 'ok', data: post});
					
					if (i == posts.length - 1) {
						done();
					}
				});
			}
		});
	});
	
	it('mysql_post.MysqlPostTest.add()', function(done) {
		let _this = this;
		
		_this.timeout(0);
		
		// 添加post
		_this._pst.add({content: 'content'}, function(err, res) {
            assert(!err);
            
			// 测试添加状态ok
			assert.equal(res.status, 'ok');
			
			// 测试添加的id合法
			let post = res.data;
			let id = post.id;
			assert(typeof id == 'number' && id > 0);

			// 测试添加的timestamp合法
			let timestamp = post.timestamp;
			assert(typeof timestamp == 'number' && timestamp > 0);
			
			// 测试添加的content正确
			let content = post.content;
			assert.equal(content, 'content');
			
			done();
		});
	});
	
	it('mysql_post.MysqlPostTest.modify()', function(done) {
		let _this = this;
		
		_this.timeout(0);
		
		// 添加测试post
		_this._pst.add({content:'content'}, function() {});
		
		// 获取post
		_this._pst.getAll(function(err, res) {
			let prePost = res.data[0];
			
			// 修改post
			_this._pst.modify({id:prePost.id, content:'CONTENT'}, function(merr, mres) {
                assert(!merr);
                
				// 测试修改状态ok
				assert.equal(mres.status, 'ok');
				
				// 保存修改后的post
				let modifiedPost = mres.data;
				
				// 再次获取post
				_this._pst.getAll(function(aerr, ares) {
					let nowPost = ares.data[0];
					
					// 测试获取的id和修改后的id一致
					assert.equal(prePost.id, modifiedPost.id);
					
					// 测试获取的timestamp和修改后的timestamp一致
					assert.equal(prePost.timestamp, modifiedPost.timestamp);
					
					// 测试修改后的content正确
					assert.equal(modifiedPost.content, 'CONTENT');
					
					// 测试修改后的post和再次获取的post一致
					assert.deepEqual(modifiedPost, nowPost);
					
					done();
				});
			});
		});
	});
	
	it('mysql_post.MysqlPostTest.remove()', function(done) {
		let _this = this;
		
		_this.timeout(0);

		// 添加post
		_this._pst.add({content: 'content'}, function() {});
		
		// 获取post ID
		_this._pst.getAll(function(err, res) {
			let id = res.data[0].id;
			
			// 删除post
			_this._pst.remove({id: id}, function(rerr, rres) {
				// 测试删除结果
                assert(!rerr);
				assert.deepEqual(rres, {status: 'ok', data: {id: id}});
				
				done();
			});
		});
	});
});

File: server/python/test/hz/message/mysql_post_test.py
#!/usr/bin/env python
#-*-coding: utf-8-*-

'''测试mysql_Post'''

import os, sys
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../../../src/'))
from hz.message.mysql_holder import MysqlHolder
from hz.message.mysql_post import MysqlPost
import mysql.connector
import time
from datetime import datetime
import unittest

# 数据库用户
_user = 'root'
# 数据库密码
_password = 'Ningning~1'
# 数据库名称
_database = 'test_message'

class MysqlPostTest(unittest.TestCase):
    '''测试mysql_holder.MysqlPost'''
    # 测试post
    __testPosts = [
        {'content': 'abcdefghijklmnopqrstuvwxyz'},
        {'content': 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'},
        {'content': '1234567890'},
        {'content': '红尘世界？一片雾茫茫！'},
        {'content': '<p>红尘世界？<em>一片雾茫茫！</em></p>'},
    ]

    def setUp(self):
        self.__holder = MysqlHolder(_user, _password, _database)
        self.__post = MysqlPost(self.__holder.inst())

    def tearDown(self):
        self.__holder.destroy()
        self.__holder.close()

    def test_getCount(self):
        '''测试mysql_post.MysqlPost.getCount()'''
        # 测试初始post个数为0
        res = self.__post.getCount()
        self.assertEqual(res, {'status': 'ok', 'data': {'count': 0}})

        # 添加测试post
        for post in self.__testPosts:
            self.__post.add(post)
            
        # 测试post个数为添加的post个数
        res = self.__post.getCount()
        self.assertEqual(res, {'status': 'ok', 'data': {'count': len(self.__testPosts)}})

    def test_getAll(self):
        '''测试mysql_post.MysqlPost.getAll()'''
        # 测试初始的post为空
        res = self.__post.getAll()
        self.assertEqual(res, {'status': 'ok', 'data': []})

        # 添加测试post
        for post in self.__testPosts:
            self.__post.add(post)

        # 测试获取的post时间递减，和添加的post顺序相反、content一致
        res = self.__post.getAll()
        self.assertEqual(res['status'], 'ok')
        posts = res['data']
        self.assertEqual(len(posts), len(self.__testPosts))
        savedTimestamp = sys.maxint
        for i in range(len(posts)):
            post = posts[i]
            timestamp = post['timestamp']
            self.assertTrue(timestamp <= savedTimestamp)
            savedTimestamp = timestamp;
            content = post['content']
            self.assertEqual(content, self.__testPosts[len(self.__testPosts) - i - 1]['content'])

    def test_getById(self):
        '''测试mysql_post.MysqlPost.getById()'''
        # 测试对空的post，通过id获取错误
        res = self.__post.getById({'id': sys.maxint})
        self.assertDictContainsSubset({'status': 'none_target'}, res)

        # 添加测试post
        for post in self.__testPosts:
            self.__post.add(post)

        # 获取所有的post，用于比对
        res = self.__post.getAll()
        posts = res['data']

        # 测试对每一个添加的post，和通过对应的id获取的post一致
        for post in posts:
            res = self.__post.getById({'id': post['id']})
            self.assertEqual(res, {'status': 'ok', 'data': post})

    def test_add(self):
        '''测试mysql_post.MysqlPost.add()'''
        # 添加post
        res = self.__post.add({'content': 'content'})
        
        # 测试添加状态ok
        self.assertDictContainsSubset({'status': 'ok'}, res)
        
        # 测试添加的id合法
        id = res['data']['id']
        self.assertTrue(id > 0)
        
        # 测试添加的timestamp合法
        timestamp = res['data']['timestamp']
        self.assertTrue(isinstance(timestamp, int) and timestamp > 0)
        
        # 测试添加的content正确
        content = res['data']['content']
        self.assertEqual(content, 'content')

    def test_modify(self):
        '''测试mysql_post.MysqlPost.modify()'''
        # 添加测试post
        self.__post.add({'content': 'content'})

        # 获取post
        res = self.__post.getAll()
        prePost = res['data'][0]

        # 修改post
        res = self.__post.modify({'id': prePost['id'], 'content': 'CONTENT'})
        
        # 测试修改状态ok
        self.assertDictContainsSubset({'status': 'ok'}, res)
        
        # 保存修改后的post
        modifiedPost = res['data']

        # 再次获取post
        res = self.__post.getAll()
        nowPost = res['data'][0]

        # 测试获取的id和修改后的id一致
        self.assertEqual(prePost['id'], modifiedPost['id'])
        
        # 测试获取的timestamp和修改后的timestamp一致
        self.assertEqual(prePost['timestamp'], modifiedPost['timestamp'])
        
        # 测试修改后的content正确
        self.assertEqual(modifiedPost['content'], 'CONTENT')
        
        # 测试修改后的post和再次获取的post一致
        self.assertEqual(modifiedPost, nowPost)

    def test_remove(self):
        '''测试mysql_post.MysqlPost.remove()'''
        # 添加post
        self.__post.add({'content': 'content'})

        # 获取post ID
        res = self.__post.getAll()
        id = res['data'][0]['id']

        # 删除post
        res = self.__post.remove({'id': id})
        
        # 测试删除结果
        self.assertEqual(res, {'status': 'ok', 'data': {'id': id}})

if __name__ == '__main__':
    unittest.main()
